aws/rust-runtime/aws-config/src/sts/assume_role.rs

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Assume credentials for a role through the AWS Security Token Service (STS).
+
+use aws_sdk_sts::error::AssumeRoleErrorKind;
+use aws_sdk_sts::{Client, Config};
+use aws_types::credentials::{self, future, CredentialsError, ProvideCredentials};
+use aws_types::region::Region;
+
+use tracing::Instrument;
+
+/// Credentials provider that uses credentials provided by another provider to assume a role
+/// through the AWS Security Token Service (STS).
+///
+/// When asked to provide credentials, this provider will first invoke the inner credentials
+/// provider to get AWS credentials for STS. Then, it will call STS to get assumed credentials for
+/// the desired role.
+///
+/// # Examples
+/// ```rust
+/// use aws_config::sts::{AssumeRoleProvider};
+/// use aws_types::{Credentials, region::Region};
+/// use aws_config::environment;
+/// use aws_config::environment::credentials::EnvironmentVariableCredentialsProvider;
+///
+/// let provider = AssumeRoleProvider::new(
+///     EnvironmentVariableCredentialsProvider::new(),
+///     "arn:aws:iam::123456789012:role/demo",
+///     Region::from_static("us-east-2"),
+///     "testAR",
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct AssumeRoleProvider<P> {
+    provider: P,
+    role: String,
+    session: String,
+    region: Region,
+}
+
+impl<P> AssumeRoleProvider<P>
+where
+    P: ProvideCredentials + Clone + 'static,
+{
+    /// Use the given `provider` to assume the given `role`.
+    ///
+    /// `role` is the Amazon Resource Name (ARN) of the role to assume.
+    /// `session` is a temporary identifier for the assumed role session.
+    pub fn new(
+        provider: P,
+        role: impl Into<String>,
+        region: Region,
+        session: impl Into<String>,
+    ) -> Self {
+        Self {
+            provider,
+            role: role.into(),
+            region,
+            session: session.into(),
+        }
+    }
+
+    async fn credentials(&self) -> credentials::Result {
+        let span = tracing::info_span!("assume_role",
```

Comment:
is it worth making an another function so we can just instrument the outer future and avoid the span dancing?

Reply:
Probably — I'll make that change.

Reply:
Turns out this is nicer with `#[tracing::instrument]`.